zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+
+        for ((&height, qblocks), expected_height) in qrange.zip(expected_heights) {
+            // Check that the heights are continuous
+            if height != expected_height {
+                // There is a gap in the block chain
+                // Missing blocks are not an error - wait for more blocks
+                return (Some(expected_height), Some(expected_hash));
+            }
+
+            // Check if any queued block at this height is part of the chain
+            // TODO(teor): consider using find_map() with a qblock method
+            let parent_hash = qblocks
+                .iter()
+                .skip_while(|qblock| qblock.hash != expected_hash)
+                .map(|qblock| qblock.block.header.previous_block_hash)
+                .next();
+
+            if let Some(parent_hash) = parent_hash {
+                expected_hash = parent_hash;
+                last_height = Some(height);
+            } else {
+                // There was no valid block in the queued block list
+                // Missing blocks are not an error - wait for more blocks
+                //
+                // This check also handles zero-length vecs, which should be
+                // impossible
+                return (Some(expected_height), Some(expected_hash));
+            }
+        }
+
+        // If the loop didn't execute, we're waiting for the next checkpoint block
+        // TODO(teor): add tests for this case
+        if last_height.is_none() {
+            return (Some(next_checkpoint_height), Some(next_checkpoint_hash));
+        }
+
+        // At this point, last_height is the height of the last block, but
+        // expected_hash is the hash of the last block's parent block.
+        //
+
+        // If we're at the genesis block, we want to return None for the parent
+        // block height.
+        let parent_height = last_height.unwrap().0.checked_sub(1).map(BlockHeight);
+        (parent_height, Some(expected_hash))
+    }
+
+    /// Check all the blocks in the current checkpoint range. Send `Ok` for the
+    /// blocks that are in the chain, and `Err` for side-chain blocks.
+    ///
+    /// Does nothing if verification has finished.
+    ///
+    /// Check that the current checkpoint range is complete before calling this
+    /// function.
+    fn submit_current_checkpoint_range(&mut self) {
+        // If checkpoint verification has finished, there should be no queued blocks.
+        if self.current_checkpoint_range.is_none() {
+            return;
+        }
+
+        let next_checkpoint_height = self.get_next_checkpoint_height().unwrap();
+        let current_range = self.current_checkpoint_range.unwrap();
+
+        // Verify all the blocks and discard all the bad blocks in the current range.
+        let mut expected_hash = self.get_next_checkpoint_hash().unwrap();
+        let qrange = self.queued.range_mut(current_range).rev();
+        for (_, qblocks) in qrange {
```

Comment:
you might be able to break this into a free function so you can just write

```rust
for (_, qblocks) in self.current_queued_range()  {
```

Reply:
Sounds like a good idea. We'll need mut and non-mut variants.

Reply:
I don't know if we'll still need this change after the range check refactor in https://github.com/ZcashFoundation/zebra/pull/524#pullrequestreview-440347176

Reply:
Most of the calls to this function were deleted as part of the rewrite.